[Composable] Add hybrid shard AC compile test

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105207
* #105206
* #105205
* #105180
* #105176

This was request to ensure hybrid shard + AC + compile works.

Differential Revision: [D47462393](https://our.internmc.facebook.com/intern/diff/D47462393/)

rohan-varma committed:
[Composable] Add hybrid shard AC compile test

This was request to ensure hybrid shard + AC + compile works.

Differential Revision: [D47462393](https://our.internmc.facebook.com/intern/diff/D47462393/)

[ghstack-poisoned]

awgu commented on test/distributed/_composable/fully_shard/test_fully_shard_compile.py:
```diff
@@ -39,9 +39,11 @@ def test_compile(self):
                 ],
                 "skip_fsdp_guards": [True, False],
-                "act_checkpoint": [True],
+                "act_checkpoint": [True, False],
```
As commented on in the previous PR, I think this change should be done in that PR.

awgu commented on test/distributed/_composable/fully_shard/test_fully_shard_compile.py:
```diff
@@ -39,9 +39,11 @@ def test_compile(self):
                     ShardingStrategy.FULL_SHARD,
                     ShardingStrategy.SHARD_GRAD_OP,
                     ShardingStrategy.NO_SHARD,
+                    ShardingStrategy.HYBRID_SHARD,
+                    ShardingStrategy._HYBRID_SHARD_ZERO2,
```
How do these get tested in CI or even with a single node locally? What process groups get constructed?

Is the replication group a singleton process group and the sharding process group takes all GPUs?

awgu replied:
(I think the answer to this affects whether we should land this PR as is, or if we need to do some changes on manually specifying PGs.)

rohan-varma replied:
that's a good point, I will check this. I ran it on a single node so I wonder if it just created an FSDP group and singel rank data parallel groups.

rohan-varma replied:
@awgu I think we just need a similar world size change as in the prev diff.

rohan-varma committed:
Update on "[Composable] Add hybrid shard AC compile test"


This was request to ensure hybrid shard + AC + compile works.

Differential Revision: [D47462393](https://our.internmc.facebook.com/intern/diff/D47462393/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Composable] Add hybrid shard AC compile test"


This was request to ensure hybrid shard + AC + compile works.

Differential Revision: [D47462393](https://our.internmc.facebook.com/intern/diff/D47462393/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Composable] Add hybrid shard AC compile test"


This was request to ensure hybrid shard + AC + compile works.

Differential Revision: [D47462393](https://our.internmc.facebook.com/intern/diff/D47462393/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Composable] Add hybrid shard AC compile test"


This was request to ensure hybrid shard + AC + compile works.

Differential Revision: [D47462393](https://our.internmc.facebook.com/intern/diff/D47462393/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Composable] Add hybrid shard AC compile test"


This was request to ensure hybrid shard + AC + compile works.

Differential Revision: [D47462393](https://our.internmc.facebook.com/intern/diff/D47462393/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Composable] Add hybrid shard AC compile test"


This was request to ensure hybrid shard + AC + compile works.

Differential Revision: [D47462393](https://our.internmc.facebook.com/intern/diff/D47462393/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI flaky"
